jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/internal/ContentSourceRange.java

```diff
@@ -0,0 +1,187 @@
+    private final boolean _readToEof;
+    private long _offsetRemaining;
+    private long _lengthRemaining;
+
+    /**
+     * Create a {@link ContentSourceRange} which wraps another {@link Content.Source} to appear
+     * as a sub-range of the original.
+     *
+     * @param source The {@link Content.Source} to wrap.
+     * @param offset the offset byte of the content to start from.
+     *               Must be greater than or equal to 0 and less than the content length (if known).
+     * @param length the length of the content to make available, -1 for the full length.
+     *               If the size of the content is known, the length may be truncated to the content size minus the offset.
+     * @throws IndexOutOfBoundsException if the offset or length are out of range.
+     * @see TypeUtil#checkOffsetLengthSize(long, long, long)
+     */
+    public ContentSourceRange(Content.Source source, long offset, long length)
+    {
+        this(source, offset, length, false);
+    }
+
+    /**
+     * Create a {@link ContentSourceRange} which wraps another {@link Content.Source} to appear
+     * as a sub-range of the original.
+     *
+     * @param source The {@link Content.Source} to wrap.
+     * @param offset the offset byte of the content to start from.
+     *               Must be greater than or equal to 0 and less than the content length (if known).
+     * @param length the length of the content to make available, -1 for the full length.
+     *               If the size of the content is known, the length may be truncated to the content size minus the offset.
+     * @param readToEof - if true, read until EOF of the source after the range is read.
+     * @throws IndexOutOfBoundsException if the offset or length are out of range.
+     * @see TypeUtil#checkOffsetLengthSize(long, long, long)
+     */
+    public ContentSourceRange(Content.Source source, long offset, long length, boolean readToEof)
+    {
+        _source = source;
+        _offset = offset;
+        _length = TypeUtil.checkOffsetLengthSize(offset, length, source.getLength());
+        _offsetRemaining = _offset;
+        _lengthRemaining = _length;
+        _readToEof = readToEof;
+    }
+
+    @Override
+    public long getLength()
+    {
+        return _length;
+    }
+
+    @Override
+    public Content.Chunk read()
+    {
+        while (true)
+        {
+            Content.Chunk chunk = _source.read();
+            if (chunk == null)
+                return null;
+
+            if (Content.Chunk.isFailure(chunk))
+                return chunk;
+
+            if (_offsetRemaining > 0)
+            {
+                if (_offsetRemaining >= chunk.remaining())
+                {
+                    // We can skip this whole chunk.
+                    _offsetRemaining -= chunk.remaining();
+                    if (chunk.isLast())
+                    {
+                        chunk.clear();
+                        return chunk;
```

Comment:
I find this confusing, and I think it is missing a call to `release()`.
I don't think we can trust the application to call `release()` on a last+empty chunk, as it is often assumed that the buffer inside is `BufferUtil.EMPTY_BUFFER` and so it does not need `release()`.
I would rather do it here where we know we have to `release()`.

Should not it be:


```suggestion
                        chunk.release();
                        return Chunk.EOF;
```